www/js/GenWidget.js

```diff
@@ -0,0 +1,158 @@
+
+  this.codeField = this.createCodeField({
+    value: 'font-size 24px;',
+    change: (e) => this._codeFieldUpdate(e)
+  })
+
+  this.slider = w.createSlider({
+    background: '#f00',
+    min: 0,
+    max: 100,
+    label: 'PX'
+    change: (e) => this._fontSliderUpdate(e)
+  })
+
+*/
+class CodeField extends HTMLElement {
+  connectedCallback (opts) {
+    this.innerHTML = `<div>
+        <button>insert</button>
+        <input type="text">
+    </div>`
+
+    this.input = this.querySelector('input')
+    this.insertButton = this.querySelector('button')
+
+    this.querySelector('button').addEventListener('click', () => {
+      const val = this.querySelector('input').value
+      this.from = NNE.cm.getCursor('from')
+      this.to = NNE.cm.getCursor('to')
+      NNE.cm.replaceSelection(val)
+      const t = { line: this.to.line, ch: this.from.ch + val.length }
+      NNE.cm.setSelection(this.from, t)
+    })
+
+    this.querySelector('input').addEventListener('input', (e) => {
+      this.setAttribute('value', e.target.value)
+      this.change(e)
+    })
+  }
+
+  get value () {
+    return this.getAttribute('value')
+  }
+
+  set value (val) {
+    this.setAttribute('value', val)
+    const input = this.querySelector('input')
+    if (input) {
+      input.value = val
+    }
+  }
+
+  static get observedAttributes () {
+    return ['value']
+  }
+
+  attributeChangedCallback (attrName, oldVal, newVal) {
+    if (newVal !== oldVal) this[attrName] = newVal
+  }
+}
+
+window.customElements.define('code-field', CodeField)
+
+class GenWidget extends Widget {
+  constructor (opts) {
+    super(opts)
+    console.log('genwidget constructed')
+  }
+
+  createCodeField (opts) {
+    const el = document.createElement('code-field')
+    el.setAttribute('value', '12')
+    el.value = 13
+    el.change = opts.change
+
+    return el
+  }
+
+  createSlider (opts) {
+    const el = document.createElement('div')
+    el.className = 'gen-slider-div'
+
+    const slider = document.createElement('input')
+    slider.setAttribute('type', 'range')
+    slider.setAttribute('min', opts.min || '1')
+    slider.setAttribute('max', opts.max || '100')
+    slider.setAttribute('step', opts.step || '1')
+    slider.setAttribute('value', opts.value || '50')
+    slider.className = 'gen-slider'
+    slider.addEventListener('input', function (e) {
+      // el.setAttribute('value', slider.value)
+      // console.log(slider.value)
+      opts.change(e)
+    })
+    const label = document.createElement('span')
+    label.innerHTML = opts.label || ''
+    label.className = 'gen-slider-label'
+
+    const bubble = document.createElement('div')
+    bubble.style = 'width: 15px;height: 15px;border-radius: 50%;border: 2px solid var(--netizen-meta);transform: translate(-7px,17px);position: relative;left: 127px;'
+    bubble.className = 'gen-slider-bubble'
+
+    el.appendChild(slider)
+    el.appendChild(label)
+    return el
```

Comment:
so say we create a new slider via the API like this:
```js
    this.fontSlider = this.createSlider({
      background: '#f00',
      min: 0,
      max: 100,
      label: 'PX',
      change: (e) => this._fontSliderUpdate(e)
    })
```

then later we want to change the slider value like this:
```js
  _codeFieldUpdate (e) {
    const css = this.parceCSS(e.target.value)
    this.fontSlider.value = window.parseInt(css.value[1])
  }
```
this won't work b/c `this.fontSlider` is a `div` element (the `el` being returned on line 122 is the parent container for the slider) && has no 'value' property. i think this is essentially the same issue/challenge we had w/the code-field, so i'm thinking we should probably create another custom element for this, maybe a 'code-slider', similar to how u did for the code-field.

as u do that, take a look at the `ColorWidget`'s `_createHTML` method, b/c at the moment the way the slider is being created isn't exactly the same as the ColorWidget sliders:
![Selection_737](https://user-images.githubusercontent.com/2506806/98481343-29483400-21bf-11eb-8af0-35dd6c24a58d.png)
notice that the label (PX) is on the wrong side, also notice that the slider isn't styled the same way